R/make-array.R

```diff
@@ -80,6 +80,137 @@ makeMR <- function (subvariables, name, selections, ...) {
+                         not_selected = "not_selected",
+                         unanswered = NA,
+                         ...) {
+    if (missing(name)) {
+        halt("Must supply a name for the new variable")
+    }
+    if (is.Categorical(var) || is.Text(var)) {
+        uniques <- names(table(var))
+    } else {
+        halt(dQuote(substitute(var)),
+             " must be a Categorical or Text Crunch Variable.")
+    }
+    items <- unique(unlist(strsplit(uniques, delim)))
+    
+    # make a derivation expression for each unique item 
+    subvarderivs <- lapply(items, function(x) createSubvarDeriv(var, x, delim,
+        selected, not_selected, unanswered))
+    names(subvarderivs) <- items
+    
+    # generate the ZCL to make an array from the subvariable derivations, and 
+    # then do selection magic to make an MR
+    derivation <- zfunc("select_categories",
+                        zfunc("array", 
+                              zfunc("select", list(map=subvarderivs),
+                                    list(value=I(c(1, 2, 3, 4, 5))))),
+                        list(value=I("selected")))
+    
+    # hide the original variable
+    var <- hide(var)
+    
+    return(VariableDefinition(derivation=derivation, name=name, ...))
+}
+
+#' Create subvariable derivation expressions
+#'
+#' This function creates a single subvariable definition based on a character string
+#' to search for and an originating variable. It uses regex to determine whether
+#' a string is present in a delimited list, then substitutes the user supplied values
+#' to indicate selection, non-selection, and missingness.
+#'
+#'
+#' @inheritParams mrFromDelim
+#' @param str A string whose presence indicates a selection
+#' @param missing A logical vector indicating which variable entries are missing
+#' @keywords internal
+#'
+#' @return A VariableDefinition
+createSubvarDeriv <- function (var, str, delim, selected, not_selected,
+                               unanswered) {
+    if (is.na(unanswered)) {
+        unanswered <- "No Data"
+    }
+    new_cat_type <- list(
+        value = list(
+            class = "categorical",
+            categories = list(
+                list("id" = 1,
+                    "name" = unanswered,
+                    "numeric_value" = NA,
+                    "missing" = TRUE),
+                list("id" = 2,
+                    "name" = selected,
+                    "numeric_value" = NA,
+                    "missing" = FALSE),
+                list("id" = 3,
+                    "name" = not_selected,
+                    "numeric_value" = NA,
+                    "missing" = FALSE)
+             )
+        )
+    )
+    new_cat <- list(column = I(1:3), type = new_cat_type)
+    deriv <- zfunc("case", new_cat)
+    deriv$args[[2]] <- zfunc("is_missing", var)
+    deriv$args[[3]] <- zfunc("~=", var, buildDelimRegex(str, delim))
+    deriv$references <- list(name = str, alias = paste0(alias(var), "_", str))
```

Comment:
We should probably check that these aliases are unique in the dataset. Currently the concatenate the source variable's alias with the item name (`str`). 

It should be doable to check that this alias doesn't exist, and if it does do something else.